datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -1520,9 +1639,79 @@ impl SMJStream {
         } else {
             self.output_size -= record_batch.num_rows();
         }
-        self.output_record_batches.clear();
+        if self.filter.is_none() || self.join_type != JoinType::Left {
+            self.output_record_batches.batches.clear();
+        }
         Ok(record_batch)
     }
+
+    fn filter_joined_batch(&mut self) -> Result<RecordBatch> {
```

Comment:
here is the main logic: But in fact what it does it just consolidate batches, filtering information and does the filtering in very similar way we currently do in `freeze_streamed`